python/ray/dataframe/dataframe.py

```diff
@@ -49,38 +46,63 @@ def _get_index(self):
         Returns:
             The union of all indexes across the partitions.
         """
-        if self._pd_index is None:
-            self._pd_index = Index.to_pandas(self._index)
-
-        return self._pd_index
+        return self._index.index
 
     def _set_index(self, new_index):
         """Set the index for this DataFrame.
 
         Args:
             new_index: The new index to set this
         """
-        self._pd_index = None
-        self._index = Index.from_pandas(new_index, self._lengths)
+        self._index.index = new_index
 
     def _default_index(self):
-        dest_indices = [(i, j)
-                        for i in range(len(self._lengths))
-                        for j in range(self._lengths[i])]
-        return Index({i: dest_indices[i] for i in range(len(dest_indices))},
-                     pd.RangeIndex)
+        """Create a default index, which is a RangeIndex
+
+        Returns:
+            The pd.RangeIndex object that represents this DataFrame.
+        """
+        dest_indices = {"partition":
+                        [i for i in range(len(self._lengths))
+                         for j in range(self._lengths[i])],
+                        "index_within_partition":
+                        [j for i in range(len(self._lengths))
+                         for j in range(self._lengths[i])]}
+        return pd.DataFrame(dest_indices)
```

Comment:
This is our index. `pd.DataFrame("partition": [...], "index_within_partition": [...])`. We use the index of this DataFrame as our index, and the destination within our DataFrame is stored in each row. 

Each partition of the DataFrame is Indexed with a RangeIndex, because it is much smaller and sometimes faster.